### PR TITLE
frontend: Allow tagged users to add external services

### PR DESCRIPTION
**Pull request by ryanslade**
This allows us to allow specific users to add external services. It could also be used to opt in a random percentage of users by running a once off query.

### CODE REVIEW

**Comment by ryanslade** [comment]
@asdine I know you had thoughts on how could do progressive rollout but once I saw that we already have the concept of user tags this seemed like an easier change

File: cmd/frontend/graphqlbackend/external_services_test.go [hunk 1]
```diff
@@ -114,6 +122,49 @@ func TestAddExternalService(t *testing.T) {
 				t.Fatalf("NamespaceUserID: want %d but got %d", userID, *result.externalService.NamespaceUserID)
 			}
 		})
+
+		t.Run("user mode disabled but user has public tag", func(t *testing.T) {
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/external_services_test.go:126]: nit: just to align with the tone with other tests :D
```suggestion
		t.Run("user mode not enabled but user has public tag", func(t *testing.T) {
```

Or change other occurrences to use "disabled".

File: cmd/frontend/graphqlbackend/external_services_test.go [hunk 2]
```diff
@@ -114,6 +122,49 @@ func TestAddExternalService(t *testing.T) {
...
+			conf.Mock(&conf.Unified{
+				SiteConfiguration: schema.SiteConfiguration{
+					ExternalServiceUserMode: "disabled",
+				},
+			})
+			defer conf.Mock(nil)
+
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/external_services_test.go:133]: It is currently only enabled if it is "public" (per RFC 211), so technically these lines are not required but doesn't hurt.

File: cmd/frontend/graphqlbackend/external_services_test.go [hunk 3]
```diff
@@ -114,6 +122,49 @@ func TestAddExternalService(t *testing.T) {
...
+			defer func() {
+				db.Mocks.ExternalServices = db.MockExternalServices{}
+			}()
+
+			oldMock := backend.MockCheckActorHasTag
+			defer func() {
+				backend.MockCheckActorHasTag = oldMock
+			}()
+
+			backend.MockCheckActorHasTag = func(context.Context, string) error { return nil }
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/external_services_test.go:146]: To make this test more useful, WDYT we mock the `db.Users.GetByID` to return a user with the tag `TagUserExternalServicePulibc` and let the `backend.CheckActorHasTag` do the real check?

**Review by unknwon — APPROVED** [review]

File: cmd/frontend/graphqlbackend/external_services_test.go (hunk 3 above)
Comment (ryanslade) [reply cmd/frontend/graphqlbackend/external_services_test.go:146]: Yeah, that sounds better, good idea.

**Pull request merged**